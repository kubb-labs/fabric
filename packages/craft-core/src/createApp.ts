import { defineApp } from './defineApp.ts'
import { ref } from './reactive/ref.ts'
import type * as KubbFile from './types'

export const createApp = defineApp<{ files: Array<KubbFile.File> }>((app) => {
  const files = ref<Array<KubbFile.File>>(app.files)
  const output = ref<string>('')

  return {
    files,
    output,
    async run() {
      throw new Error('Method not implemented')
    },
    async waitUntilExit() {
      throw new Error('Method not implemented')
    },
  }
})
