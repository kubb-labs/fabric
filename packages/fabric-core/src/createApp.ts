import { defineApp } from './defineApp.ts'

export const createApp = defineApp()
