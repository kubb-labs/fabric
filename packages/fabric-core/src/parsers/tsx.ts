import { typeScriptParser } from './typescript.ts'
import { createParser } from './createParser.ts'

export const tsxParser = createParser({
  name: 'tsx',
  install() {},
  async print(file, options = { extname: '.tsx' }) {
    return typeScriptParser.print(file, options)
  },
})
