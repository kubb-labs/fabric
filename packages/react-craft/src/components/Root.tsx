import { Component, createContext } from 'react'

import type { KubbNode } from '../types.ts'

type ErrorBoundaryProps<Meta extends Record<string, unknown> = Record<string, unknown>> = {
  onError: (error: Error) => void
  meta: Meta
  children?: KubbNode
}

class ErrorBoundary extends Component<{
  onError: ErrorBoundaryProps['onError']
  children?: KubbNode
}> {
  state = { hasError: false }

  static displayName = 'KubbErrorBoundary'
  static getDerivedStateFromError(_error: Error) {
    return { hasError: true }
  }

  componentDidCatch(error: Error) {
    if (error) {
      this.props.onError(error)
    }
  }

  render() {
    if (this.state.hasError) {
      return null
    }
    return this.props.children
  }
}

export type RootContextProps<Meta extends Record<string, unknown> = Record<string, unknown>> = {
  /**
   * Exit (unmount) the whole Ink app.
   */
  readonly exit: (error?: Error) => void
  readonly meta: Meta
}

export const RootContext = createContext<RootContextProps>({
  exit: () => {},
  meta: {},
})

type RootProps<Meta extends Record<string, unknown> = Record<string, unknown>> = {
  /**
   * Exit (unmount) hook
   */
  readonly onExit: (error?: Error) => void
  /**
   * Error hook
   */
  readonly onError: (error: Error) => void
  readonly meta: Meta
  readonly children?: KubbNode
}

export function Root<Meta extends Record<string, unknown> = Record<string, unknown>>({ onError, onExit, meta, children }: RootProps<Meta>) {
  try {
    return (
      <ErrorBoundary
        onError={(error) => {
          onError(error)
        }}
      >
        <RootContext.Provider value={{ meta, exit: onExit }}>{children}</RootContext.Provider>
      </ErrorBoundary>
    )
  } catch (_e) {
    return null
  }
}

Root.Context = RootContext
Root.displayName = 'KubbRoot'
