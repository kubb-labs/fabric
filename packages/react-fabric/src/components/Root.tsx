import { Component, createContext } from 'react'

import type { KubbNode } from '../types.ts'

type ErrorBoundaryProps = {
  onError: (error: Error) => void
  children?: KubbNode
}

class ErrorBoundary extends Component<{
  onError: ErrorBoundaryProps['onError']
  children?: KubbNode
}> {
  state = { hasError: false }

  static displayName = 'KubbErrorBoundary'
  static getDerivedStateFromError(_error: Error) {
    return { hasError: true }
  }

  componentDidCatch(error: Error) {
    if (error) {
      this.props.onError(error)
    }
  }

  render() {
    if (this.state.hasError) {
      return null
    }
    return this.props.children
  }
}

export type RootContextProps = {
  /**
   * Exit (unmount) the whole Ink app.
   */
  readonly exit: (error?: Error) => void
}

export const RootContext = createContext<RootContextProps>({
  exit: () => {},
})

type RootProps = {
  /**
   * Exit (unmount) hook
   */
  readonly onExit: (error?: Error) => void
  /**
   * Error hook
   */
  readonly onError: (error: Error) => void
  readonly children?: KubbNode
}

export function Root({ onError, onExit, children }: RootProps) {
  try {
    return (
      <ErrorBoundary
        onError={(error) => {
          onError(error)
        }}
      >
        <RootContext.Provider value={{ exit: onExit }}>{children}</RootContext.Provider>
      </ErrorBoundary>
    )
  } catch (_e) {
    return null
  }
}

Root.Context = RootContext
Root.displayName = 'KubbRoot'
