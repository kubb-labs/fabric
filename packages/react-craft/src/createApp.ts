import { defineApp, type KubbFile, ref } from '@kubb/craft-core'
import type { ElementType } from 'react'
import { createElement } from './index.ts'
import { ReactTemplate } from './ReactTemplate.tsx'

export const createApp = defineApp<ElementType>((container) => {
  const files = ref<Array<KubbFile.File>>([])
  const output = ref<string>('')
  const template = new ReactTemplate({ files, output })

  return {
    files,
    output,
    mount() {
      return template.render(createElement(container))
    },
    waitUntilExit() {
      return template.waitUntilExit()
    },
  }
})
