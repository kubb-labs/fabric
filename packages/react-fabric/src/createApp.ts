import { defineApp } from '@kubb/fabric-core'
import { type ElementType, createElement } from 'react'
import { type ReactAppContext, ReactTemplate } from './ReactTemplate.tsx'

export const createApp = defineApp<ElementType, ReactAppContext>((app, context) => {
  const template = new ReactTemplate(context)

  return {
    render() {
      template.render(createElement(app))
    },
    async renderToString() {
      return template.renderToString(createElement(app))
    },
    waitUntilExit() {
      return template.waitUntilExit()
    },
  }
})
