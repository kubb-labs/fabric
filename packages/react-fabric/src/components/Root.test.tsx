import { vi } from 'vitest'
import { createApp } from '../createApp.ts'
import { Root } from './Root.tsx'

function Thrower(): React.ReactNode {
  throw new Error('boom')
}

describe('<Root/>', () => {
  test('render Root with children', async () => {
    const Component = () => {
      return (
        <Root onExit={() => {}} onError={() => {}}>
          Hello from Root
        </Root>
      )
    }

    const app = createApp(Component)
    const output = await app.renderToString()

    expect(output).toMatchSnapshot()
  })

  test('error boundary should catch and call onError', async () => {
    const onError = vi.fn()
    const onExit = vi.fn()

    const Component = () => {
      return (
        <Root onExit={onExit} onError={onError}>
          <Thrower />
        </Root>
      )
    }

    const app = createApp(Component)
    const output = await app.renderToString()

    expect(output).toMatchSnapshot()
    expect(onError).toHaveBeenCalledTimes(1)
    expect(onExit).not.toHaveBeenCalled()
  })
})
