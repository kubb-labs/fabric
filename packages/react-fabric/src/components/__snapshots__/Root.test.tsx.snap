// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Root/> > error boundary should catch and call onError 1`] = `""`;

exports[`<Root/> > render Root with children 1`] = `"Hello from Root"`;
